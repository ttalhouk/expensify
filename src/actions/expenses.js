import uuid from 'uuid';
import database from '../firebase/firebase';


export const addExpense = (expense) => {
  return {
    type: 'ADD_EXPENSE',
    expense
  };
}

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = {
      description,
      note,
      amount,
      createdAt
    }
    return database.ref('expenses').push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    }).catch((e) => {
      console.log(e);
    });
  };
};

export const removeExpense = ({id}) => {
  return {
    type:'REMOVE_EXPENSE',
    id
  }
}

export const startRemoveExpense = ({id}) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).remove().then(() => {
      dispatch(removeExpense({id}));
    }).catch((e) => {
      console.log(e);
    });
  };
};

export const editExpense = (id, updates) => {
  return (
    {
      type: 'EDIT_EXPENSE',
      id,
      updates
    }
  );
};
export const startEditExpense = (id, updates) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).update(updates).then(() => {
      dispatch(editExpense(id, updates));
    }).catch((e) => {
      console.log(e);
    });
  };
};

export const setExpenses = (expenses) => {
  return {
    type: 'SET_EXPENSES',
    expenses
  }
}

export const startSetExpenses = () => {
  return (dispatch) => {
    return database.ref('expenses').once('value')
      .then((snapshot) => {
        const expenses = [];
        snapshot.forEach((expense) => {
            expenses.push({
              id: expense.key,
              ...expense.val()
            })
          });
        dispatch(setExpenses(expenses));
      }).catch((error) => {
        console.log(error);
      });
  }
}
