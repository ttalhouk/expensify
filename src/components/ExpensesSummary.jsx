import React from 'react';
// import PropTypes from 'prop-types'
import { connect } from 'react-redux';
import numeral from 'numeral'

import selectExpenses from '../selectors/expenses';
import totalExpenses from '../selectors/expense-total';

export const ExpensesSummary = ({expenses}) => {
  return(
    <div>
      <h3>You are viewing {expenses.length} {expenses.length === 1 ? 'expense' : 'expenses'} totaling {numeral(totalExpenses(expenses) / 100).format('$0,0.00')}</h3>
    </div>
  )
}

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  }
}

export default connect(mapStateToProps)(ExpensesSummary);
