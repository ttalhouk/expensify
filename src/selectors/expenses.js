import moment from 'moment';

const getVisibleExpenses = (expenses, {text, sortBy, startDate, endDate}) => {
  return expenses.filter((expense) => {
    // const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate;
    // const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate;
    const startDateMatch = startDate ? startDate.isSameOrBefore(moment(expense.createdAt), 'day') : true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(moment(expense.createdAt), 'day') : true;
    const textMatch =  expense.description.toLowerCase().includes(text.toLowerCase());

    return textMatch && startDateMatch && endDateMatch;
  }).sort((exp1, exp2) => {
    if (sortBy === 'date') {
      return exp2.createdAt - exp1.createdAt;
    } else if (sortBy === 'amount'){
      return exp2.amount - exp1.amount;
    }
  })
}

export default getVisibleExpenses;
