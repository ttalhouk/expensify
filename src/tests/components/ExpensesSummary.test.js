import React from 'react';
import {shallow} from 'enzyme';

import {ExpensesSummary} from '../../components/ExpensesSummary.jsx';

import {expenses} from '../fixtures/expenses';

describe('<ExpensesSummary />', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<ExpensesSummary expenses={expenses} />);
  });
  it('should display the summary of multiple filtered items correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('should display the summary of one filtered item correctly', () => {
    wrapper.setProps({expenses: [expenses[0]]});
    expect(wrapper).toMatchSnapshot();
  });
})
