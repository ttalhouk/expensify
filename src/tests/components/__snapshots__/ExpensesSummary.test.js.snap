// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ExpensesSummary /> should display the summary of multiple filtered items correctly 1`] = `
<div>
  <h3>
    You are viewing 
    3
     
    expenses
     totaling 
    $5.90
  </h3>
</div>
`;

exports[`<ExpensesSummary /> should display the summary of one filtered item correctly 1`] = `
<div>
  <h3>
    You are viewing 
    1
     
    expense
     totaling 
    $0.50
  </h3>
</div>
`;
